core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -572,20 +572,35 @@ private[spark] class ExecutorAllocationManager(
   }
 
   /**
+   * Callback invoked when an extra executor is needed (See SPARK-19326)
+   */
+  private def onExtraExecutorNeeded(): Unit = synchronized {
+    val maxNeeded = math.max(math.min(maxNumExecutorsNeeded + 1, maxNumExecutors), minNumExecutors)
```

Comment:
we cannot just ensure there are more than two active executors left. it will depends on if there is any speculative jobs not launched.

hostToLocalTaskCount will make the new executors request satisfies the locality requirement. Also, on the other hand, if even the new executor was not guaranteed to be different host, it will idle and then die. and the speculative task will request another executor.